File: kotlin/src/commonMain/kotlin/lib/humanoid/paromat/node/Nodes.kt
package lib.humanoid.paromat.node

sealed interface Node

class NumberNode(
    val number: Double
): Node {
    override fun toString() = number.toString()
}

sealed class UnaryNode(
    val operator: Char,
    val node: Node
): Node {
    override fun toString() = "$operator($node)"
}

sealed class BinaryNode(
    val left: Node,
    val operator: Char,
    val right: Node
): Node {
    override fun toString() = "$left $operator $right"
}

class PlusNode(node: Node): UnaryNode('+', node)
class MinusNode(node: Node): UnaryNode('-', node)

class AddNode(left: Node, right: Node): BinaryNode(left, '+', right)
class SubtractNode(left: Node, right: Node): BinaryNode(left, '-', right)
class MultiplyNode(left: Node, right: Node): BinaryNode(left, '*', right)
class DivideNode(left: Node, right: Node): BinaryNode(left, '/', right)

File: python/paromat/nodes.py
class Node:
    def __init__(self):
        self.left: Node = None
        self.right: Node = None

    def __repr__(self):
        return f"{self.left} {self.right}"
    
class NumberNode(Node):
    def __init__(self, value: float):
        self.value = value

    def __repr__(self):
        return str(self.value)

class OperatorNode(Node):
    def __init__(self, operator: str, left: Node, right: Node):
        self.operator = operator
        self.left = left
        self.right = right

    def __repr__(self):
        return f"({self.left} {self.operator} {self.right})"

class AddNode(OperatorNode):
    def __init__(self, left, right):
        super().__init__('+', left, right)

class SubtractNode(OperatorNode):
    def __init__(self, left, right):
        super().__init__('-', left, right)

class MultiplyNode(OperatorNode):
    def __init__(self, left, right):
        super().__init__('*', left, right)

class DivideNode(OperatorNode):
    def __init__(self, left, right):
        super().__init__('/', left, right)

class UnaryNode(Node):
    def __init__(self, operator: str, right: NumberNode):
        self.operator = operator
        self.right = right

    def __repr__(self):
        return f"{self.operator}{self.right}"

class PlusNode(UnaryNode):
    def __init__(self, right: Node):
        self.operator = '+'
        self.right = right

class MinusNode(UnaryNode):
    def __init__(self, right: Node):
        self.operator = '-'
        self.right = right